lib/fluent/plugin/out_forward/load_balancer.rb

```diff
@@ -0,0 +1,102 @@
+#
+# Fluentd
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+require 'fluent/output'
+
+module Fluent::Plugin
+  class ForwardOutput < Output
+    class LoadBalancer
+      class NoNodesAvailable < StandardError; end
+
+      def initialize(log)
+        @log = log
+        @weight_array = []
+        @rand_seed = Random.new.seed
+        @rr = 0
+      end
+
+      def select_healthy_node
+        error = nil
+
+        wlen = @weight_array.length
+        wlen.times do
+          @rr = (@rr + 1) % wlen
+          node = @weight_array[@rr]
+          next unless node.available?
+
+          begin
+            ret = yield node
+            return ret, node
+          rescue
+            # for load balancing during detecting crashed servers
+            error = $!  # use the latest error
+          end
+        end
+
+        raise error if error
+        raise NoNodesAvailable, "no nodes are available"
```

Comment:
How about using `ForwardOutput::NoNodesAvailable`?

```
modle Fluent::Plugin
  class ForwardOutput < Output
    # error classes
  end
end

require 'fluent/plugin/out_forward/load_balancer'
```

or extract classes into `fluent/plugin/out_forward/error.rb`

Reply:
i prefer the latter. https://github.com/fluent/fluentd/pull/2483/commits/f9d50bbb457361174ebf9b148b2a85a7a08b2d7a